### ISSUE TITLE
RFE: Switch to use a Night Polar for jib/genoa instead of adjusting polar eff at Dusk.

### ISSUE THREAD

**Issue by ozolli** [rgleason/weather_routing_pi#97]
There are some options which I often use on Sailgrib and I would like to see them implemented here:

- Polar efficiency below 10kn TWS
- Polar efficiency above 20kn TWS
- Night time polar loss of efficiency

This is particularly useful with new or few crew onboard, or in case of conservative cruising.

**Comment by rgleason** [comment]
Good ideas that might fit well in the  Configuration > Advanced Tab.

**Comment by ozolli** [comment]
Ah I forgot another important setting:
When the boat speed is less than a given value, then motor at a given speed.

**Comment by rgleason** [comment]
It would also help to have a new constraint added. 
A point that would force routing to port or starboard
See https://github.com/rgleason/weather_routing_pi/issues/84

**Comment by ozolli** [comment]
This is what Saulgrib does but a different way.
On Sailgrib you first make a route with any number of waypoints, then on each waypoint you can set to go to the waypoint or to let it on port or on starboard.

Then you let Sailgrib calculate the routing with this route and its constraints.

After the routing is calculated you can add a pivot point constraint anywhere on the isochrones and have it recalculated until you get what you want:

[https://www.youtube.com/watch?v=4IkREAwjc-M](url)

(Set English subtitles).

**Comment by rgleason** [comment]
Yes, this is where I saw the feature.

**Comment by rgleason** [comment]
![Edit-config-Adjust-Polar-Eff](https://github.com/rgleason/weather_routing_pi/assets/2363959/9f947f62-6bad-4731-8078-370c5970a596)

We need to make the mathematical relationship between 
Wind Strength  X%
and 
Polar efficiency
 < __kn TWS  _%
 > __kn TWS  _%
Night time polar loss of efficiency  - __%

Clear.

**Comment by sebastien-rosset** [comment]
Some programs also have the ability to configure a day versus night efficiency. It's a very useful feature because it's quite common to sail more conservatively during the night.

**Comment by rgleason** [comment]
Yes that would take it out of the realm of guesswork.

Also see this comment about  SailGribWR and port/starboard marks constraints. i'd forgotten about Ozzoli's comment.  https://github.com/rgleason/weather_routing_pi/issues/97#issuecomment-1165400125

**Comment by ozolli** [comment]
> Some programs also have the ability to configure a day versus night efficiency. It's a very useful feature because it's quite common to sail more conservatively during the night.

The simple way to do this, and I guess they all do it this way, is to apply a polar percentage.

This can be okay but I believe there is a better way closest to reality : Use another polar at night.
A lot of people don't use a kite/spinnaker/code at night so applying a percentage on a polar with a kite is useless. The TWA are not the same as a with a genoa/jib used at night.
So a jib/genoa only polar at night is the way to go.

**Comment by rgleason** [comment]
Oliver, I like your idea, it has great merit, using the system we already have, and think we should change this RFE and perhaps close it after I update the Manual with the idea.

**Comment by sebastien-rosset** [comment]
> > Some programs also have the ability to configure a day versus night efficiency. It's a very useful feature because it's quite common to sail more conservatively during the night.
> 
> The simple way to do this, and I guess they all do it this way, is to apply a polar percentage.
> 
> This can be okay but I believe there is a better way closest to reality : Use another polar at night.

To implement this in the WR plugin, we would allow the user to specify two sets of polars, one for day time and the other for night time. Or alternatively, have one set and tag which polars are used for day time and night time.
Then during the calculations, for each isochron use the polar based on the sunrise/sunset times at the local coordinates. In principle this concept could be extended to other conditions that trigger switching polars.
This is certainly doable.

We might want to support both percentages and multi polars.
1. Use separate polars when you are willing to customize polars to achieve better accuracy.
2. Use percentages for a simple experience. Ie no need to customize a polar for night time.

There is also a third approach which is do nothing and let the user manually create legs for day time and night time, but that requires a lot of guesswork and is very tedious.

> A lot of people don't use a kite/spinnaker/code at night so applying a percentage on a polar with a kite is useless.

I agree for the first part but I would not conclude the percentage feature would be useless. Some sailors do use spinnakers at night depending on the exact conditions (for example you might have a lower tolerance for TWD and swell).
Or you may have legs that are mostly upwind anyway. 

> The TWA are not the same as a with a genoa/jib.
> So a jib/genoa only polar at night is the way to go.

I think as mentioned above providing both options could be useful. If the user wants a simple option and/or they are coming from a different software that supports percentages, then they will probably miss that feature. If they want higher accuracy and are willing to spend time customizing the polars, then sure they could do that too .

That reminds me a feature I merged recently. Previously, if the wind was very light and below the first TWD column in the polar, then the WR would stop the propagation which could potentially cause the entire calculations to fail. The workaround was to create another polar with smaller TWD values. This is ok in principle, but IMO routing should work with out of the box polars from the manufacturer. So now when facing very light winds, the WR interpolates the STW based on the polar value. This provides a simpler and less frustrating experience especially for first time users. Then over time these users can learn the more advanced scenarios with polar customization.

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset  Oliver, I like your idea, it has great merit, using the system we already have, and think we should change this RFE and perhaps close it after I update the Manual with the idea.
> 
> However, when doing an automatic routing over many days, we would need some day/night trigger using dawn/dusk time to automatically switch to and from the night polar file!!!
> 
> I guess having some polar efficiency value, would also require an automatic switch on/off at dusk/dawn,  but using a special nighttime jib/genoa polar is probably more accurate.

Yes I was planning to compare the isochron time with sunrise/sunset times and apply day time vs night time percentages. The same capability could be used for night time polar switching. With both approaches, some work is required.

**Comment by sebastien-rosset** [comment]
> Polar efficiency below 10kn TWS
> Polar efficiency above 20kn TWD
> Night time polar loss of efficiency

I'm trying to think how the first two options would be implemented. Somebody else might argue well for me the threshold is 23 knots.  Or I need 3 different thresholds, 10, 20 and 30.
And similarly during night time there could be different thresholds.
Potentially this could even be connected to the crew schedule, though that would make the feature significantly more complicated.

I understand the value of night time versus day time efficiency. But is it really necessary to support multiple TWD thresholds?

**Comment by ozolli** [comment]
> I understand the value of night time versus day time efficiency. But is it really necessary to support multiple TWD thresholds?

I don't think so.
The simple (and better) option is to set a global efficiency percentage, then cumulative night efficiency percentage.

That would mean, for example, a polar speed of 10kn, a global 90% and a night 80% would mean 9kn at day and 7.2kn at night.

qtVlm use upwind %, downwind %, and cumulative night %.
They don't have night polar so when I'm routing I set a mix of downwind and night % to mimic a "no kite at night" polar.

**Comment by sebastien-rosset** [comment]
I propose to add the following parameters:
1. **Upwind Efficiency**: Applied during day sailing when on upwind points of sail (default: 100%)
2. **Downwind Efficiency**: Applied during day sailing when on downwind points of sail (default: 100%)
3. **Night Cumulative Efficiency**: Additional multiplier applied at night on top of the upwind/downwind parameters (default: 100%)

Example calculation:

Polar speed: 10kn
If upwind during day with 90% efficiency: 9kn
If downwind during day with 85% efficiency: 8.5kn
If upwind during night with 90% day efficiency and 80% night efficiency: 7.2kn
If downwind during night with 85% day efficiency and 80% night efficiency: 6.8kn

Open Item:
I'm still debating if these parameters should be saved with the boat (e.g. Boat.xml) or for each route configuration.

I'm leaning towards adding these parameters in the RouteMapConfiguration, and I'm open to suggestions.

**Comment by rgleason** [comment]
I haven't thought about this much, but why wouldn't it become a part of the default settings that are found under the General and Advanced Tabs?

**Comment by sebastien-rosset** [comment]
> I haven't thought about this much, but why wouldn't it become a part of the default settings that are found under the General and Advanced Tabs?

It seems the philosophy of the plugin is that all routing parameters are either part of the boat configuration or the routing configuration. Some lucky sailors have more than one boat, or they may want to adjust the settings for each routing based on crew or other parameters. Sometimes I put a lower value as a quick way to take tidal current into consideration.

BTW, I think there should be a global setting to specify the default boat to use for new routing configurations, but still each routing can be configured with a different boat.

**Comment by rgleason** [comment]
Now home. Need to publish a new version and try these.

**Comment by rgleason** [comment]
@ozolli 
@sebastien-rosset 

I've tested WR 1.15.21.21 using several different values for Upwind, Downwind and Nightime.
I set all to 100 and ran a routing starting at night.
I set Upwind to 90, Downwind to 80 and night to 50 and ran it starting at 8 pm.  I noticed a significant increase in time.
I set Upwind to 90, Downwind to 80 and night to 50  changed it to Daytime 8am and it was faster.
I set all to 100 and ran a routing starting at daytime.  and it was even faster.

Relevant Fact: Believe it or not, the air increased in density enough between summer and spring/fall to make a 5-7% difference in force.   Should there be some temperature adjustment?  ---Just kidding.

**Comment by rgleason** [comment]
Closing.
